pkg/kubelet/cm/devicemanager/pod_devices.go

```diff
@@ -321,10 +328,70 @@ func (pdev *podDevices) deviceRunContainerOptions(podUID, contName string) *Devi
 			annotationsMap[k] = v
 			opts.Annotations = append(opts.Annotations, kubecontainer.Annotation{Name: k, Value: v})
 		}
+
+		// Updates for CDI devices. This also updates the annotations.
+		cdiDevices, cdiAnnotations := getCDIDeviceInfo(
+			podUID, resourceName,
+			resp.CdiDevices,
+			annotationsMap,
+			cdiDevicesMap,
+		)
+		opts.CDIDevices = append(opts.CDIDevices, cdiDevices...)
+		opts.Annotations = append(opts.Annotations, cdiAnnotations...)
```

Comment:
Do we need to dedup the annotations here? Is it possible that someone has set the annotations exactly the same as what `getCDIDeviceInfo()` will set. If so, does it matter if they are duplicated in the `opts.Annotations` slice?

**Edit**, I see you do that below with:
```
			klog.V(4).InfoS("Skip existing annotation", "annotationKey", a.Name, "annotationValue", a.Value)
			if e != a.Value {
				klog.ErrorS(nil, "Annotation has conflicting setting", "annotationKey", a.Name, "expected", e, "got", a.Value)
			}
			continue
```

Can we pull the deduplication up to this level rather than burying it inside `getCDIDeviceInfo()` I feel like it would be clearer that way. It would also clean up the api to  `getCDIDeviceInfo()` since you would now drop the following 2:
```
			annotationsMap,
			cdiDevicesMap,
````
(since I'm assuming the only reason you passed an empty cdiDevicesMap in in the first place was to make the API look more symmetrical).

Reply:
I think the likelihood of collisions in the annotations is pretty low, but since we can't rule them out entirely, I have left the warnings / errors in.

I don't agree that the `cdiDevicesMap` is there just to make the API symetrical. We also deduplicate the CDI devices requested at this level which requires the device map. Note that it's only empty at the first call after which it will contain all CDI devices for the first ContainerAllocateResponse processed.

Reply:
Ah, right -- because its in a loop.

Reply:
If deduplication is happening in both cases, then keeping it inside the `getCDIDeviceInfo` call makes sense.